sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ComplexTypes.scala

```diff
@@ -59,6 +63,13 @@ object SimplifyExtractValueOps extends Rule[LogicalPlan] {
           Literal(null, ga.dataType)
         }
       case GetMapValue(CreateMap(elems), key) => CaseKeyWhen(key, elems)
+      case GetMapValue(Literal(map: MapData, MapType(kt, vt, _)), key) =>
+        val elems = new mutable.ListBuffer[Literal]
+        map.foreach(kt, vt, (key, value) => {

```

Comment:
Don't bother unless you're changing it otherwise, but `.foreach { case (kt, vt, (key, value)) =>`?
Is the ListBuffer there for efficiency vs just flatMap-ing to two elements for every map element?
```
val result = map.flatMap { case ((kt, vt, (key, value)) => Seq(Literal(key, kt), Literal(value, vt)) }
```

Reply:
well, `map` is a `MapData` object, which has a `foreach`, but no `flatMap` at all, so what you are suggesting is just not feasible, unless we add the flatMap method in `MapData`, but that is an overkill IMO

Reply:
Ah right, disregard that.